Source/WebKit/Platform/IPC/Connection.cpp

```diff
@@ -342,9 +369,10 @@ void Connection::setShouldExitOnSyncMessageSendFailure(bool shouldExitOnSyncMess
 // IPC messages as some messages may get received on the IPC thread before the message receiver registered itself on the main thread.
 void Connection::enqueueMatchingMessagesToMessageReceiveQueue(MessageReceiveQueue& receiveQueue, const ReceiverMatcher& receiverMatcher)
 {
-    ASSERT(isMainRunLoop());
+    assertIsCurrent(m_dispatcher);
```

Comment:
For these sequential access assertions, you could do:

```
Locker lock { m_sequentialAccess };

Connection {
  NO_UNIQUE_ADDRESS SequetialAccessLock;
}
```

This could be done in a separate patch that also introduces WTF::SequentialAccessLock. (The lock is just a dummy that asserts that threadid == locked thread id.)

Reply:
Isn't that a considerably weaker guarantee than the current assertions?

A consumer could use a single connection from multiple threads (without external synchronisation), and depending on timing might not hit the assertions during testing. The code would still be broken though, and could cause issues with different workloads/timings.

What's the benefit of allowing this 'non-sequential' pattern, rather than being permanently associated with a given thread (or SerialFunctionDispatcher thing)?

Reply:
> Isn't that a considerably weaker guarantee than the current assertions?

Yes. In general we want the weakest possible guarantees that satisfies the requirements of the system. This gives the widest possible uses and widest possible future modification directions.

So we need to meticulously understand what the system requires.

> A consumer could use a single connection from multiple threads (without external synchronisation), and depending on timing might not hit the assertions during testing. The code would still be broken though, and could cause issues with different workloads/timings.

This is a valid argument.
 
> What's the benefit of allowing this 'non-sequential' pattern, rather than being permanently associated with a given thread (or SerialFunctionDispatcher thing)?

I explained this in the other comment.
Unless needed, I'd be hesitant to add the semantics of "SerialFunctionDispatcher either is current or is not current". This is because then all SerialFunctionDispatchers would need that property to apply to the functions dispatched through them. This would (possibly?) mean that the functions dispatched through them would need to carry this additional state even in cases it would not be carried.

If you think it'd be easier, maybe it's beneficial to add this to FunctionDispatcher / SerialFunctionDispatcher and be done with it, and then I can think if it's really a problem later.